### PR TITLE
Normalize rollback nodes when a transaction is constructed

### PR DESCRIPTION
**Pull request by nickchapman-da**
Done:
- Implemented normalization rules as new code pass in `NormalizeRollbacks`
- Fairly extensive tests in `NormalizeRollbacksSpec`
- ensure pre-order traversal of node-ids is in increasing order; and starts from 0, fixes some existing tests which depend on this
- make normalization pass stack-safe (CPS & trampolines)
- enable normalization pass in `PartialTransaction`

part of #8020

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+          // normalization rule #2 or #3
+          List(roll)
+
+        case Case.Single(act: Norm.Act) =>
+          // no rule
+          List(Norm.Roll1(act))
+
+        case Case.Multi(h: Norm.Roll, m, t) =>
+          // normalization rule #2
+          h :: makeRoll(m ++ List(t))
```
Comment (sofiafaro-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:99]: This recursion here should also be made stack-safe.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+class NormalizeRollbacksSpec extends AnyWordSpec with Matchers with Inside {
+
+  import NormalizeRollbackSpec._
+
+  // TODO: https://github.com/digital-asset/daml/issues/8020
+  // Below we test a set of hand-constructed testcases. Hopefully we got all edge cases!
+  // But for more confidence we might consider adding scalacheck based testing.
+
+  // TODO: https://github.com/digital-asset/daml/issues/8020
+  // We should test that the `meaning` of a transaction is preserved by normalization.
```
Comment (sofiafaro-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala:28]: Good idea. "Meaning" here would be the transaction tree among actions (i.e. after removing/flattening all rollback nodes), as well as the consistency order and the set of rolled back actions. That's a lot, but we definitely want to guarantee we're preserving these.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+  // normalization rules #2 and #3
+  test("mixed-rollback-nesting-A")(
+    Top(R(R(c1), c2, R(c3))),
+    Top(R(c1), R(c2, c3)),
+  )
+  test("mixed-rollback-nesting-B")(
+    Top(R(R(c1), c2, c3, R(c4))),
+    Top(R(c1), R(c2, c3, c4)),
+  )
+
```
Comment (sofiafaro-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala:158]: These tests look great! I would also add some more odd mixed rollback cases:

```
R(c1, R(R(c2), c3)) --> R(c1, R(c2), c3)
```

and 

```
R(R(c1, R(c2)), c3) --> R(c1, c2), R(c3)
```

**Review by sofiafaro-da — COMMENTED** [review]
This normalization algorithm looks right, and the unit tests are extensive. Thank you!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+    import Canonical.{Norm, Case, caseNorms}
+    // The normalization phase works by constructing intermediate values which are
+    // `Canonical` in the sense that only properly normalized nodes can be represented.
+    // Although this doesn't ensure correctness, one class of bugs is avoided.
+
+    // The `force*` functions move from the world of Norms into standard transactions.
+    // This is also where node-ids for the resulting normalized transaction are generated.
+
+    def forceRoll(x: Norm.Roll): Node = {
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:57]: what exactly are we forcing here? I don’t understand the naming csheme.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+      final case object Empty extends Case
+      final case class Single(n: Norm) extends Case
+      final case class Multi(h: Norm, m: List[Norm], t: Norm) extends Case
+    }
+
+    def caseNorms(xs: List[Norm]): Case = {
+      xs.length match {
+        case 0 => Case.Empty
+        case 1 => Case.Single(xs(0))
+        case n => Case.Multi(xs(0), xs.slice(1, n - 1), xs(n - 1))
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:184]: Maybe worth using a `Vector` or an `ImmArray` instead of slicing & calling length on a single-linked list? @remyhaemmerle-da probably has a better idea what makes sense here.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    // The List is necessary because the rules can:
+    // (1) drop nodes; (2) combine nodes (3) lift nodes from a lower level to a higher level.
+
+    // TODO: https://github.com/digital-asset/daml/issues/8020
+    // make this traversal code stack-safe
+
+    txOriginal match {
+      case GenTransaction(nodesOriginal, rootsOriginal) =>
+        def traverseNids(nids: List[Nid]): List[Norm] = nids.flatMap(traverseNid)
+        def traverseNid(nid: Nid): List[Norm] = traverseNode(nodesOriginal(nid))
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:130]: this doesn’t look stacksafe afaict.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+        def traverseNode(node: Node): List[Norm] = {
+          node match {
+
+            case NodeRollback(children) =>
+              makeRoll(traverseNids(children.toList))
+
+            case ex: NodeExercises[_, _] =>
+              val norms = traverseNids(ex.children.toList)
+              val children = pushNodes(forceNorms(norms))
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:140]: From what I can tell this might be where you’re breaking the node id order. Now children of an exercise will have a lower node id than their parent. As mentioned on slack, I think we need to preserve node ids = index in preorder at least for Daml Studio and at that point it’s probably easiest to preserve it everywhere (and it’s also useful for debugging over the ledger API).

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    // `Canonical` in the sense that only properly normalized nodes can be represented.
+    // Although this doesn't ensure correctness, one class of bugs is avoided.
+
+    // The `force*` functions move from the world of Norms into standard transactions.
+    // This is also where node-ids for the resulting normalized transaction are generated.
+
+    def forceRoll(x: Norm.Roll): Node = {
+      x match {
+        case Norm.Roll1(act) =>
+          val child = pushNode(act.node)
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:60]: the logic around when we call `forceRoll` definitely looks like it doesn’t result in `pushNode` calls in prefix order either.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+  // Shape: description of a transaction, with conversions to and from a real tx
+  sealed trait Shape
+  object Shape {
+
+    final case class Top(xs: List[Shape])
+    final case class Create(x: Long) extends Shape
+    final case class Exercise(x: List[Shape]) extends Shape
+    final case class Rollback(x: List[Shape]) extends Shape
+
+    def toTransaction(top: Top): TX = {
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala:212]: this looks like we can probably simplify the code a fair bit by using the transaction builder.

**Review by cocreature — COMMENTED** [review]
Nice work!
I think there are two main points we need to address before merging:
1. Node ids should be indices in preorder traversal. While this is not required for a valid transaction, this is very useful for users and we shouldn’t break it here.
2. We should fix the code to be stacksafe

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala]: done

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala (hunk 3 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala:169]: added this, thanks

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 4 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala]: renamed `force*` to `push*` and updated explanatory comment

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 6 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala]: everything should be stack safe now

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 7 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala]: the node ordering is fixed now, and also tested

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 8 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala]: fixed

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala (hunk 9 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala:255]: I looked at the transaction builder, and it look really awkward to use in this case.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:252]: calling length on List is probably a good idea. 
Why not using standard List pattern matching ?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+      final case object Empty extends Case
+      final case class Single(n: Norm) extends Case
+      final case class Multi(h: Norm, m: List[Norm], t: Norm) extends Case
+    }
+
+    def caseNorms(xs: List[Norm]): Case = {
+      xs.length match {
+        case 0 => Case.Empty
+        case 1 => Case.Single(xs(0))
+        case n => Case.Multi(xs(0), xs.slice(1, n - 1), xs(n - 1))
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:252]:
```suggestion
      xs match {
        case Nil => Case.Empty
        case x0 :: Nil => Case.Single(h)
        case x0 :: x1 :: tail  => Case.Multi(x0, List(x1), tail)
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+      def run: A = {
+        this match {
+          case Land(a) => a
+          case Bounce(func) => func().run
+        }
+      }
+    }
+    final case class Land[A](a: A) extends Tramp[A]
+    final case class Bounce[A](func: () => Tramp[A]) extends Tramp[A]
+  }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:270]: Could we factorize the Trampoline of `com.daml.lf.speedy.Anf` ? 
I will put it in `com.daml.lf.data`

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+import scala.annotation.tailrec
+
+private[lf] object NormalizeRollbacks {
+
+  type Nid = NodeId
+  type Cid = Value.ContractId
+  type TX = GenTransaction[Nid, Cid]
+  type Node = GenNode[Nid, Cid]
+  type LeafNode = LeafOnlyActionNode[Cid]
+  type ExeNode = NodeExercises[Nid, Cid]
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:21]: Why this is not private ? 

```suggestion
  private[this] type Nid = NodeId
  private[this]  type Cid = Value.ContractId
  private[this] type TX = GenTransaction[Nid, Cid]
  private[this] type Node = GenNode[Nid, Cid]
  private[this] type LeafNode = LeafOnlyActionNode[Cid]
  private[this] type ExeNode = NodeExercises[Nid, Cid]
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+  }
+
+  // makeRoll: encodes the normalization transformation rules:
+  //   rule #1: ROLL [ ] -> ε
+  //   rule #2: ROLL [ ROLL [ xs… ] , ys… ] -> ROLL [ xs… ] , ROLL [ ys… ]
+  //   rule #3: ROLL [ xs… , ROLL [ ys… ] ] ->  ROLL [ xs… , ys… ]
+
+  //   rule #2/#3 overlap: ROLL [ ROLL [ xs… ] ] -> ROLL [ xs… ]
+
+  private def makeRoll[R](norms: List[Norm])(k: List[Norm] => Tramp[R]): Tramp[R] = {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:95]:
```suggestion
  private[this] def makeRoll[R](norms: List[Norm])(k: List[Norm] => Tramp[R]): Tramp[R] = {
```
almost only cosmetic

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+      case Case.Multi(h: Norm.Act, m, t: Norm.Act) =>
+        // no rule
+        k(List(Norm.Roll2(h, m, t)))
+    }
+  }
+
+  private def pushIntoRoll(a1: Norm.Act, xs2: List[Norm], t: Norm.Roll): Norm.Roll = {
+    t match {
+      case Norm.Roll1(a3) => Norm.Roll2(a1, xs2, a3)
+      case Norm.Roll2(a3, xs4, a5) => Norm.Roll2(a1, xs2 ++ List(a3) ++ xs4, a5)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:128]:
```suggestion
      case Norm.Roll2(a3, xs4, a5) => Norm.Roll2(a1, xs2 ++ a3 +: xs4, a5)
```
Anyway is concatenating lists a good idea ? 
Either add a comment to explain why this is not linear, either you use better data structure (e.g. Frontstack, Backstack, Vector ...)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+    }
+  }
+
+  // State manages a counter for node-id generation, and accumulates the nodes-map for
+  // the normalized transaction
+
+  // There is no connection between the ids in the original and normalized transaction.
+
+  private case class State(index: Int, nodeMap: Map[Nid, Node]) {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:137]:
```suggestion
  private[this] case class State(index: Int, nodeMap: Map[Nid, Node]) {
```
:)

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala (hunk 9 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala:255]: Just a bit curious about what is awkward.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 10 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:252]: That’s not the same. Before we had first, middle, last. Now you have first, second, rest.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+  }
+
+  private def pushNorm[R](s: State, x: Norm)(k: (State, Nid) => Tramp[R]): Tramp[R] = {
+    x match {
+      case act: Norm.Act => pushAct(s, act)(k)
+      case roll: Norm.Roll => pushRoll(s, roll)(k)
+    }
+  }
+
+  private def pushNorms[R](s: State, xs: List[Norm])(
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:201]: I find the trampolined CPS style here pretty hard to follow. I hacked up a version of this that doesn’t use cps or trampolines and at least for me seems much easier to follow (ignore variable names and so on) https://github.com/digital-asset/daml/commit/2efd36a6f4efe69edbc2f48fd347cbd9b5ac916b. wdyt @remyhaemmerle-da @nickchapman-da? We can probably do something very similar for the first phase as well.

Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:201]: I think the nice thing about this approach is that it's a systematic transformation of the original code.
See:
https://github.com/digital-asset/daml/pull/9712/commits/4511623e529c9c3d022538f3bf078ce6916d214b
https://github.com/digital-asset/daml/pull/9712/commits/e520a935e6943023d0f165b450f71c7f63077335

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 10 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:252]: indeed.
We should then use `ImmArray` or `Vector`.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 11 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala]: done

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 12 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala]: done

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 13 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala]: done

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 15 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala]: done

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala (hunk 9 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala:255]: Mainly that I am building the transaction bottom up from a declarative description.
But the transaction builder seems to assume that I am building top-down. Create nodes, and then mutably attach children.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 16 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:202]: We agreed to proceed with the trampoline approach, so long as Remy is happy. Which he is!

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:202]: I am fine with both approaches.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala (hunk 5 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:253]: switched to Vector to address this concern

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Trampoline.scala [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+
+package com.daml.lf
+package data
+
+import scala.annotation.tailrec
+
+// Trampolines for stack-safety
+object Trampoline {
+
+  sealed trait Trampoline[A] {
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Trampoline.scala]: done

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Trampoline.scala [hunk 18]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package data
+
+import scala.annotation.tailrec
+
+// Trampolines for stack-safety
+object Trampoline {
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Trampoline.scala]: done
